### PR TITLE
Performance/Restore optimization

### PR DESCRIPTION
**Pull request by carljohnsen**
This pull request rewrites the restore operation to follow the same Communicating Sequential Processes structure like the backup operation.

The old restore operation is still available through the `--restore-legacy=true` option.

Whereas the old version downloaded a volume and wrote all of the target blocks within the volume, this new approach restores one file at a time, although concurrently. The volumes and the blocks are then stored in memory, if they are required by multiple files, and are auto evicted from cache when they are no longer needed.

### CODE REVIEW

**Comment by ts678** [comment]
> The volumes and the blocks are then stored in memory

[Dblock size recommendations?](https://forum.duplicati.com/t/dblock-size-recommendations/19521/2?u=ts678)

> For LAN setups, I think somewhere in the range of 100 MiB to 5 GiB would make sense.

I bit my tongue over the 5 GiB advice (thinking of storage and transfer speed), but won't that plus this fill memory?

Were any analyses, simulations, or extended tests done to see how well this does compared to the current scheme?

The reason I say extended tests is because source files might initially write into few volumes, but compacts happen.
Because there is no mechanism to keep all the blocks for a file together, eventually I would think blocks spread out.

[Blockset hash collisions?](https://forum.duplicati.com/t/blockset-hash-collisions/11477/11?u=ts678) discusses the best and worst cases of dblock packing. Even slightly worse may fill memory.

**Comment by carljohnsen** [comment]
@ts678 Ah, I can see how the text is misleading - The blocks are stored in memory (default size of 1 MB), while the volume readers are stored in memory (default size 50 MB) - i.e. the volumes themselves are kept on disk and only a reference is in memory (at least that's the intention). A block is only kept cached in memory if it's used by multiple restore target files and it's evicted from cache once it has been used by all of the target files. The same goes for the volumes, which are auto evicted once all of it's blocks have been read. These strategies should help keep memory consumption low. Another strategy would be to tune the number of file processors, which can be done by setting the `--restore-file-processors` option. Furthermore, the size of the blocks stored in cache can be adjusted by the `--restore-cache-max` option.

For a backup of >3000 files totaling >60 GB, I've seen memory consumption lie in the range of 300 - 400 MB. However, I have to confirm this through more extensive profiling, as I haven't done so yet. 

A more thorough description of the new restore process will also be added once it's been finalized. Thank you for the links, they'll help me construct edge-case tests!

**Comment by ts678** [comment]
Thanks for clarifying. I feel better about memory now, and worse about disk. Some tmp are small, but tempdir can bypass that.
True edge case might be hard to hit. Blocksize default used to be 100 KB. I don't know if dblock has any hardcoded upper limit. Typically people don't go out of their way to do insane things, but some want big backups, and performance can then be slow.
[Added tool to generate test data #5260](https://github.com/duplicati/duplicati/pull/5260) or similar might be useful for tests. Speed is always good, but some systems are small.

File: Duplicati/Library/Main/Operation/RestoreHandler.cs [hunk 1]
```diff
@@ -105,56 +108,44 @@ public void Run(string[] paths, Library.Utility.IFilter filter = null)
...
+                    RecreateDatabaseHandler.NumberedFilterFilelistDelegate filelistfilter = FilterNumberedFilelist(m_options.Time, m_options.Version);
+                    db = new LocalRestoreDatabase(tmpdb);
+                    m_result.RecreateDatabaseResults = new RecreateDatabaseResults(m_result);
+                    using (new Logging.Timer(LOGTAG, "RecreateTempDbForRestore", "Recreate temporary database for restore"))
+                        new RecreateDatabaseHandler(m_backendurl, m_options, (RecreateDatabaseResults)m_result.RecreateDatabaseResults)
+                            .DoRun(db, false, filter, filelistfilter, null);
+
+                    if (!m_options.SkipMetadata)
+                        ApplyStoredMetadata(m_options, new RestoreHandlerMetadataStorage());
+                }
```
Comment (kenkendk) [inline Duplicati/Library/Main/Operation/RestoreHandler.cs:137]: I think this comment is important.
From what I recall, this is a hack from the FE, when restoring without a pre-existing local database.
Maybe the fix is to change the FE to work correctly in this case.

**Review by kenkendk — APPROVED** [review]
Looks good!
Only critical thing is the possible version-hack that I need to investigate.

File: Duplicati/Library/Main/Options.cs [hunk 2]
```diff
@@ -88,6 +88,36 @@ public class Options
...
+
+        /// <summary>
+        /// The default value for the ratio of the restore cache to evict when full
+        /// </summary>
+        private const float DEFAULT_RESTORE_CACHE_EVICT = 0.5f;
+
+        /// <summary>
+        /// The default value for the number of file processors during restore
+        /// </summary>
+        private const int DEFAULT_RESTORE_FILE_PROCESSORS = 8;
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Options.cs:104]: Yes!

File: Duplicati/Library/Main/Options.cs [hunk 3]
```diff
@@ -421,6 +451,16 @@ public IList<ICommandLineArgument> SupportedCommands
                     new CommandLineArgument("secret-provider-pattern", CommandLineArgument.ArgumentType.String, Strings.Options.SecretProviderPatternShort, Strings.Options.SecretProviderPatternLong, SecretProviderHelper.DEFAULT_PATTERN),
                     new CommandLineArgument("secret-provider-cache", CommandLineArgument.ArgumentType.Enumeration, Strings.Options.SecretProviderCacheShort, Strings.Options.SecretProviderCacheLong, Enum.GetName(SecretProviderHelper.CachingLevel.None), null, Enum.GetNames(typeof(SecretProviderHelper.CachingLevel))),
                     new CommandLineArgument("cpu-intensity", CommandLineArgument.ArgumentType.Integer, Strings.Options.CPUIntensityShort, Strings.Options.CPUIntensityLong, "10", null, ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10"]),
+
+                    new CommandLineArgument("restore-cache-max", CommandLineArgument.ArgumentType.Size, Strings.Options.RestoreCacheMaxShort, Strings.Options.RestoreCacheMaxLong, "8G"),
+                    new CommandLineArgument("restore-cache-evict", CommandLineArgument.ArgumentType.Decimal, Strings.Options.RestoreCacheEvictShort, Strings.Options.RestoreCacheEvictLong, "0.5"),
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Options.cs:456]: Sure! Yes, percentage should be sufficient.

File: Duplicati/Library/Main/Options.cs [hunk 4]
```diff
@@ -421,6 +451,16 @@ public IList<ICommandLineArgument> SupportedCommands
                     new CommandLineArgument("secret-provider-pattern", CommandLineArgument.ArgumentType.String, Strings.Options.SecretProviderPatternShort, Strings.Options.SecretProviderPatternLong, SecretProviderHelper.DEFAULT_PATTERN),
                     new CommandLineArgument("secret-provider-cache", CommandLineArgument.ArgumentType.Enumeration, Strings.Options.SecretProviderCacheShort, Strings.Options.SecretProviderCacheLong, Enum.GetName(SecretProviderHelper.CachingLevel.None), null, Enum.GetNames(typeof(SecretProviderHelper.CachingLevel))),
                     new CommandLineArgument("cpu-intensity", CommandLineArgument.ArgumentType.Integer, Strings.Options.CPUIntensityShort, Strings.Options.CPUIntensityLong, "10", null, ["1", "2", "3", "4", "5", "6", "7", "8", "9", "10"]),
+
+                    new CommandLineArgument("restore-cache-max", CommandLineArgument.ArgumentType.Size, Strings.Options.RestoreCacheMaxShort, Strings.Options.RestoreCacheMaxLong, "8G"),
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Options.cs:455]: Alright, changed! I've also added the capability of setting the max to 0, which disables the cache.

File: Duplicati/Library/Main/Database/LocalRestoreDatabase.cs [hunk 5]
```diff
@@ -957,6 +964,114 @@ public IEnumerable<IFileToRestore> GetFilesToRestore(bool onlyNonVerified)
...
+
+        /// <summary>
+        /// Returns a list of <see cref="BlockRequest"/> for the given blockset ID. It is used by the <see cref="FileProcessor"/> to restore the blocks of a file.
+        /// </summary>
+        /// <param name="blocksetID">The BlocksetID of the file.</param>
+        /// <returns>A list of <see cref="BlockRequest"/> needed to restore the given file.</returns>
+        public IEnumerable<BlockRequest> GetBlocksFromFile(long blocksetID)
+        {
+            using var cmd = m_connection.CreateCommand();
+            cmd.CommandText = @$"
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Database/LocalRestoreDatabase.cs:1020]: I tried to do this with a prepared command, but SQLite doesn't like multiple Tasks getting readers from the same command concurrently. I'll investigate whether there's some way to do this with multiple tasks.

File: Duplicati/Library/Main/Database/LocalRestoreDatabase.cs [hunk 6]
```diff
@@ -957,6 +964,114 @@ public IEnumerable<IFileToRestore> GetFilesToRestore(bool onlyNonVerified)
...
+                SELECT Block.ID, Block.Hash, Block.Size, Block.VolumeID
+                FROM BlocksetEntry INNER JOIN Block
+                ON BlocksetEntry.BlockID = Block.ID
+                WHERE BlocksetEntry.BlocksetID = ?";
+            cmd.AddParameter();
+            cmd.SetParameterValue(0, blocksetID);
+            using var reader = cmd.ExecuteReader();
+            for (long i = 0; reader.Read(); i++)
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Database/LocalRestoreDatabase.cs:1028]: What should be a `var`?

File: Duplicati/Library/Main/Database/LocalRestoreDatabase.cs [hunk 7]
```diff
@@ -957,6 +964,114 @@ public IEnumerable<IFileToRestore> GetFilesToRestore(bool onlyNonVerified)
...
+                FROM ""{m_tempfiletable}""
+                INNER JOIN Metadataset ON ""{m_tempfiletable}"".MetadataID = Metadataset.ID
+                INNER JOIN BlocksetEntry ON Metadataset.BlocksetID = BlocksetEntry.BlocksetID
+                INNER JOIN Block ON BlocksetEntry.BlockID = Block.ID
+                WHERE ""{m_tempfiletable}"".ID = ?
+            ";
+            cmd.AddParameter();
+            cmd.SetParameterValue(0, fileID);
+            using var reader = cmd.ExecuteReader();
+            for (long i = 0; reader.Read(); i++)
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Database/LocalRestoreDatabase.cs:1053]: What should be a `var`?

File: Duplicati/Library/Main/ResultClasses.cs [hunk 8]
```diff
@@ -637,6 +637,14 @@ public override ParsedResultType ParsedResult
 
     internal class RestoreResults : BasicResults, Library.Interface.IRestoreResults
     {
+        /// <summary>
+        /// The list of broken local files - i.e. locally stored files that raised an error during restore.
+        /// </summary>
+        public List<string> BrokenLocalFiles { get; internal set; } = [];
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/ResultClasses.cs:643]: Right, changed!

File: Duplicati/Library/Main/Operation/Restore/BlockManager.cs [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+            /// Decrements the block counters for the block and volume, and checks
+            /// if the block is still needed. If the block is no longer needed, it
+            /// will be removed from the cache. If the volume is no longer needed,
+            /// the VolumeDownloader will be notified.
+            /// </summary>
+            /// <param name="blockRequest">The block request to check.</param>
+            /// <param name="decrement"></param>
+            public void CheckCounts(BlockRequest blockRequest)
+            {
+                lock (m_blockcount)
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Operation/Restore/BlockManager.cs:136]: Right! The logging has been moved outside of the lock.

File: Duplicati/Library/Main/Operation/Restore/BlockManager.cs [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+
+            /// <summary>
+            /// Retire the dictionary. This will decrement the number of readers
+            /// accessing the dictionary, and if there are no more readers, it
+            /// will retire the volume request channel, effectively shutting
+            /// down the restore process network.
+            /// </summary>
+            public void Retire()
+            {
+                if (Interlocked.Decrement(ref readers) <= 0) {
```
[the remark this answers is not in the document]
Comment (carljohnsen) [reply Duplicati/Library/Main/Operation/Restore/BlockManager.cs:239]: Right, if it only had a single channel. This is due to each `FileProcessor` having it's own request/response channel set, so that each request is returned to the respective `FileProcessor`. So this count variable is used to keep track of how many readers are still active, to avoid premature shutdown of the network.

File: Duplicati/Library/Main/Operation/RestoreHandler.cs (hunk 1 above)
Comment (carljohnsen) [reply Duplicati/Library/Main/Operation/RestoreHandler.cs:137]: Right, this shouldn't have been removed. Reintroduced the comment and the code when the database doesn't exist, which is where the code and comment resided before.
